src/main/java/ch/fhnw/algd1/converters/utf8/UTF8Converter.java

```diff
@@ -3,37 +3,167 @@
+                //intToBinaryString does not give leading "0" chars.
+                // so we do not need to trim it
+                return binaryStringToInt(mainBinaryString);
+
+            } else if (countBytes == 2) {
+                System.out.println("2 bytes");
+                String mainBinaryString = intToBinaryString(bytes[0]);
+                String followUp1 = intToBinaryString(bytes[1]);
+
+                mainBinaryString = mainBinaryString.replaceFirst("110", "");
+                followUp1 = followUp1.replaceFirst("10", "");
+
+                return binaryStringToInt(mainBinaryString + followUp1);
+
+            } else if (countBytes == 3) {
+                System.out.println("3 bytes");
+                String mainBinaryString = intToBinaryString(bytes[0]);
+                String followUp1 = intToBinaryString(bytes[1]);
+                String followUp2 = intToBinaryString(bytes[2]);
+
+                mainBinaryString = mainBinaryString.replaceFirst("1110", "");
+                followUp1 = followUp1.replaceFirst("10", "");
+                followUp2 = followUp2.replaceFirst("10", "");
+
+                return binaryStringToInt(mainBinaryString + followUp1 + followUp2);
+            } else if (countBytes == 4) {
+                System.out.println("4 bytes");
+
+                String mainBinaryString = intToBinaryString(bytes[0]);
+                String followUp1 = intToBinaryString(bytes[1]);
+                String followUp2 = intToBinaryString(bytes[2]);
+                String followUp3 = intToBinaryString(bytes[3]);
+
+                mainBinaryString = mainBinaryString.replaceFirst("11110", "");
+                followUp1 = followUp1.replaceFirst("10", "");
+                followUp2 = followUp2.replaceFirst("10", "");
+                followUp3 = followUp3.replaceFirst("10", "");
+
+                return binaryStringToInt(mainBinaryString + followUp1 + followUp2 + followUp3);
+            }
+
+            return 0;
+        } else {
+            return 0;
+        }
+    }
+
+    private static boolean isValidUTF8(byte[] bytes) {
+        if (bytes.length == 1) return (bytes[0] & 0b1000_0000) == 0;
+        else if (bytes.length == 2) return ((bytes[0] & 0b1110_0000) == 0b1100_0000) && isFollowup(bytes[1]);
+        else if (bytes.length == 3)
+            return ((bytes[0] & 0b1111_0000) == 0b1110_0000) && isFollowup(bytes[1]) && isFollowup(bytes[2]);
+        else if (bytes.length == 4)
+            return ((bytes[0] & 0b1111_1000) == 0b1111_0000) && isFollowup(bytes[1]) && isFollowup(bytes[2]) && isFollowup(bytes[3]);
+        else return false;
+    }
+
+    private static boolean isFollowup(byte b) {
+        return (b & 0b1100_0000) == 0b1000_0000;
+    }
+
+    private static int binaryStringToInt(String binary) {
+        int x = 0;
+        for (int i = 0; i < binary.length(); i++) {
+            x = (x << 1) + (binary.charAt(i) == '0' ? 0 : 1);
```

Comment:
Wenn man für die Multiplikation mit 2 die Schiebe-Operation verwendet, wäre es eigentlich konsequent, anstelle einer Addition die bitweise oder-Operation zu verwenden.
Die Fall-Unterscheidung mit dem Conditional-Operator können Sie ja inzwischen auch vermeiden, indem Sie einfach schreiben:
```suggestion
            x = (x << 1) | (binary.charAt(i) - '0');
```